Add Java-Client-API to Katacoda Scenario

mdodell opened the pull request.

mdodell commented:
- [x] Convert the java-client-api workflow to use Kubernetes instead of Openshift
- [x] Upload the java-client-api assets onto Katacoda
- [x] Redesign the workflow of the Java Client API installer to be initialized via a DockerFile
- [x] Convert every command from the installer script to go be used in an Ubuntu environment
- [x] Separate the scripts into clean markdown files to be used in the tutorial

A link to this Katacoda Scenario can be found here: [Java-Client-API Scenario](https://katacoda.com/mdodell/scenarios/java-client-api-tutorial)

izgeri commented:
@mdodell I had some thoughts after seeing you demo this scenario. I’m not sure yet which things we should try to fix before the PR is merged and which we should do as follow-ons, but if any of these are simple enough to start with I’d suggest making these changes as separate commits from what you have so far.
1. I don’t think we should include the alternate “run one script and all the magic happens” flow. It doesn’t add anything to the demo and it’s kind of confusing.
1. Maybe even more important, do you feel clear about the 4 steps of the demo? This is a different format than the original demo, so how did it translate? In this format, we’re trying to teach someone something - do our 4 steps of the demo accomplish that? If not, how would you change them?
1. Is there a way to cut down on the amount of time the tutorial takes (besides having custom base images, which we don’t have right now)? 20 minutes is a long time for a tutorial flow. Maybe after Bradley’s work is in Step 1 (building the client) can be totally dropped?
1. Are there sane defaults we can set for things so that fewer steps are required but the demo is still clear? For example, we set the authenticator ID as an env var and then interpolate it into a policy. But maybe it'd be clearer just to hardcode it and show / explain the policy that includes it?

I might have other ideas as well, but I think now that you've migrated the content and gotten it working end-to-end, the next step is to look at it with a critical eye and ask:
- can this be simplified (both in terms of content and time spent)?
- is the overall tutorial conveying the message I want to convey?
- am I explaining the purpose of all of these steps clearly enough?

You can use that you are newer to Conjur development to your advantage - are any of the steps unclear to you? Then we should probably dig into that and improve the step / enhance the explanation. The team is ready to help with whatever questions you have :)

izgeri commented:
some notes from running through it just now:
## page 1
- [x] I don't think we call it the "java client API" anywhere, and the wording sounds odd to me. can we call it the "java API" or (better) the "java client library"? it's really the java client library for the conjur API. sometimes we call it the "java SDK", but that's not quite right either.
- [x] is it typical that you have to wait for the katacoda tutorial to load in the files? that seems odd. are our files particularly big or something?
- [x] can we rename `java-client-api` to `conjur-java-client`?
- [x] for the java build (which we eventually want to remove, right?) can we run `pushd conjur-java-client` and then end on `popd` so the next page starts from the right directory?

## page 2
- [x] can we remove the `PROJECT_NAME` env var entirely and just use the hardcoded `demo-java-client` (or maybe even rename to `java-demo`?) wherever it was?
- [x] can we add some more context about what the data key does in the "create a data key" step?
- [x] why does the numbering on this page restart at (1)?
- [x] in the helm install step, are there custom values set in the custom-values.yaml? could we look at those and explain them? can we explain in this step what the helm chart is deploying?
- [ ] we can't run the "Create a Conjur account and generate an API Key" step until the conjur pods are up. how do they handle this in the quick start? can a similar approach be used here?
- [x] in "create a conjur CLI pod" why are we doing that? what's the benefit of the CLI pod?
- [ ] in the last step, "check pod status", what am I looking for? maybe this is covered by "expected output", but I think it'd be good to add some info about why this step is part of the tutorial

## page 3
- [x] in step (2) the `conjur init` command has the ability to accept flags - would using flags make this command clearer? it's pretty fuzzy right now
- [x] we should try to clean up the errors / warnings around `Unable to use a TTY - input is not a terminal or the right kind of file` and `stty: 'standard input': Inappropriate ioctl for device`
- [ ] rather than saving the username `admin` in the "authninput` file you could just pass `-u admin` on the `conjur authn login` request (which is a more standard flow anyway)
- [x] loading the policies is like black magic right now - I don't know what the policies say or why I'm loading them
- [x] why do we have two separate certificate configmaps?

## page 4
- [x] what is the java client docker image? maybe this should be explained on page 1
- [x] any manipulation of the docker image should happen at the same time as its created (eg tagging) - on this last page, I _think_ what we're trying to do is deploy the app and verify it works, and that should be the focus
- [x] I couldn't finish because the init container didn't come up, so I ended on the second to last step 😢 

I put a lot of feedback in the list above, but basically it still reads as not very clear about what the purpose of the tutorial is and how the steps along the way are helping me to achieve that purpose. mostly right now it feels like running a bunch of steps, and I'm not sure why or what I'm doing. 

in this tutorial, you are teaching!! so start out by describing what's in this lesson, the steps that you'll take to get there, and then make sure that each page of the tutorial is walking through one of those steps

I don't think it would be bad to have a main page that just sets the context and runs no commands

I would suggest that the flow look something like this:
page 1: describe the purpose of the tutorial and what steps you'll be working through (maybe run `minikube start` at the very beginning because it takes a minute and why not)
page 2: build a docker image for a simple java app (that does what?) that will use our conjur java client to connect with conjur given an access token provided by our kubernetes authenticator client sidecar
page 3: deploy conjur using our helm chart, which deploys (list of stuff). configure conjur to enable the kubernetes authenticator
page 4: configure conjur to enable the app: load policy to grant the app a host identity, add secrets the app will need, add entitlements to grant the app access to the secrets
page 5: deploy the application with the conjur authenticator client sidecar and verify it can retrieve secrets from conjur as expected

mdodell commented:
> some notes from running through it just now:
> 
> ## page 1
> * [x]  I don't think we call it the "java client API" anywhere, and the wording sounds odd to me. can we call it the "java API" or (better) the "java client library"? it's really the java client library for the conjur API. sometimes we call it the "java SDK", but that's not quite right either.
> * [x]  is it typical that you have to wait for the katacoda tutorial to load in the files? that seems odd. are our files particularly big or something?
> * [x]  can we rename `java-client-api` to `conjur-java-client`?
> * [x]  for the java build (which we eventually want to remove, right?) can we run `pushd conjur-java-client` and then end on `popd` so the next page starts from the right directory?
> 
> ## page 2
> * [x]  can we remove the `PROJECT_NAME` env var entirely and just use the hardcoded `demo-java-client` (or maybe even rename to `java-demo`?) wherever it was?
> * [ ]  can we add some more context about what the data key does in the "create a data key" step?
> * [x]  why does the numbering on this page restart at (1)?

Fixed, Katacoda just hadn't repopulated the new code yet.
> * [ ]  in the helm install step, are there custom values set in the custom-values.yaml? could we look at those and explain them? can we explain in this step what the helm chart is deploying?
> * [ ]  we can't run the "Create a Conjur account and generate an API Key" step until the conjur pods are up. how do they handle this in the quick start? can a similar approach be used here?

See the comment about the `wait-for-pod.sh` sugestion.

> * [ ]  in "create a conjur CLI pod" why are we doing that? what's the benefit of the CLI pod?
> * [ ]  in the last step, "check pod status", what am I looking for? maybe this is covered by "expected output", but I think it'd be good to add some info about why this step is part of the tutorial

*wait-for-pod.sh suggestion:*
The build script had a loop that checked the pod status every two seconds, and ran for a total of 100 seconds. If all of the pods were running, it would break from the loop. Since we are working outside of a script, I thought that running `kubectl get pods` would be satisfactory. Another option is to have the user run a short script, something along the lines of `wait-for-pod.sh --pod-name <pod-name> --num-of-contains <number-container>` that the user executes, and it prints out a message every iteration of the loop saying:
Waiting for the ____ pod to be up...
Waiting for the ____ pod to be up...
Waiting for the ____ pod to be up...
Pod is up!
> 
> ## page 3
> * [ ]  in step (2) the `conjur init` command has the ability to accept flags - would using flags make this command clearer? it's pretty fuzzy right now
> * [ ]  we should try to clean up the errors / warnings around `Unable to use a TTY - input is not a terminal or the right kind of file` and `stty: 'standard input': Inappropriate ioctl for device`

This happens in the quick start tutorial too. I think it's because of how the structure of that command is built. If you have the user say simply `conjur init`, then they can type in `yes`, `$ACCOUNT_NAME`, and `y` themselves. 

> * [ ]  rather than saving the username `admin` in the "authninput`file you could just pass`-u admin`on the`conjur authn login` request (which is a more standard flow anyway)
> * [ ]  loading the policies is like black magic right now - I don't know what the policies say or why I'm loading them
> * [ ]  why do we have two separate certificate configmaps?
> 
> ## page 4
> * [ ]  what is the java client docker image? maybe this should be explained on page 1
> * [ ]  any manipulation of the docker image should happen at the same time as its created (eg tagging) - on this last page, I _think_ what we're trying to do is deploy the app and verify it works, and that should be the focus
> * [ ]  I couldn't finish because the init container didn't come up, so I ended on the second to last step 😢
> 
> I put a lot of feedback in the list above, but basically it still reads as not very clear about what the purpose of the tutorial is and how the steps along the way are helping me to achieve that purpose. mostly right now it feels like running a bunch of steps, and I'm not sure why or what I'm doing.
> 
> in this tutorial, you are teaching!! so start out by describing what's in this lesson, the steps that you'll take to get there, and then make sure that each page of the tutorial is walking through one of those steps
> 
> I don't think it would be bad to have a main page that just sets the context and runs no commands
> 
> I would suggest that the flow look something like this:
> page 1: describe the purpose of the tutorial and what steps you'll be working through (maybe run `minikube start` at the very beginning because it takes a minute and why not)

Should page 1 be considered a step? There is an `intro.md` file, that describes the purpose of the tutorial to some extent, but I can write some clarifications there. That intro file can also be removed so the user is immediately prompted into step 1.

The `minikube start` command can also be immediately loaded in so the user does not even have to run the command. It will still take a minute for it to finish, but at least they do not have to be concerned with executing the command itself.

> page 2: build a docker image for a simple java app (that does what?) that will use our conjur java client to connect with conjur given an access token provided by our kubernetes authenticator client sidecar
> page 3: deploy conjur using our helm chart, which deploys (list of stuff). configure conjur to enable the kubernetes authenticator
> page 4: configure conjur to enable the app: load policy to grant the app a host identity, add secrets the app will need, add entitlements to grant the app access to the secrets
> page 5: deploy the application with the conjur authenticator client sidecar and verify it can retrieve secrets from conjur as expected

izgeri commented:
Thanks - I'm seeing now that my big batch of comments is hard to go back and forth on :) I'll spend some time tomorrow to migrate outstanding ones to line-level comments to make the back-and-forth easier. We may also want to hop on a short call tomorrow, because my sense is that some of the context of my questions is lost in the deluge. I'll get in touch to set that up early tomorrow!

izgeri commented on java-client-api-tutorial/assets/conjur-java-api-example.yaml:
```diff
@@ -0,0 +1,98 @@
+    spec:
+      serviceAccountName: default
+      containers:
+      - image: cyberark/conjur-kubernetes-authenticator:latest
```
```suggestion
      - image: cyberark/conjur-authn-k8s-client:latest
```

mdodell commented:
@izgeri 
**Follow-up from today's' conversation:**
- [x] Described user personas, and broke out each step into one of three personas (Infra Engineer, Security Admin, Application Developer)
- [x] Improved the flow by adding an introduction for each step, along with a conclusion at the end of each step letting the end-user know what the next step in the tutorial was
- [x] Moved the build of the Java client to the end, so all of the 'Application Developer' steps were in the same place rather than starting with building the application, but not actually running it until the end
- [x] Set minikube to begin with a start-up script
- [x] Improved the `conjur init` command using flags
- [x] Fixed the double ConfigMaps to just use one. 
- [x] Added a conclusion at the end, with links to more resources

All other improvements have been marked on your original comment.

 **Clarifications/help needed**

1. When we are waiting for pods to be online, should we create a script that prints messages that say "loading..." in the shell? I improved the description of those pods to say: "Wait for the ____ pods to run" by running the command `kubectl get pods`
2. For your suggestion regarding `authn login` under _page 3_, I'm having a lot of difficulty trying to implement them. 

For instance, for the `authn login` command, here is the output when I try either copying and pasting or even manually writing the API key (the password) in:

```
$ kubectl exec -it "$CONJUR_CLI_POD" conjur authn login -- -u admin
Please enter admin's password (it will not be echoed):
Unable to authenticate with Conjur. Please check your credentials.
command terminated with exit code 1
```
I looked in the documentation, and didn't see a flag for password. I even tried passing in $API_KEY, which is the environment variable for the API key. 

Any advice? I feel like having a password flag there would be super helpful if it doesn't exist!

izgeri commented on java-client-api-tutorial/README.md:
```diff
@@ -0,0 +1,91 @@
+# Conjur Installation Demo on OpenShift
```
can this folder be renamed?

also, this README title isn't accurate, right? should this file be deleted?

izgeri commented on java-client-api-tutorial/assets/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+# =================== BUILD API CONTAINER ===================
```
this container's purpose is to build the conjur java client .jar file, can the heading be clear about that?

izgeri commented on java-client-api-tutorial/assets/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+RUN mvn install -DskipTests -Dmaven.javadoc.skip=true
+
+# =================== MAIN CONTAINER ===================
+FROM maven:3-openjdk-8 as java-client
```
this container isn't the "java-client" container, it's a simple java app that uses the java client

izgeri commented on java-client-api-tutorial/assets/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+
+RUN mvn install -Dconjur-api-version="$VERSION"
+
+ENTRYPOINT ["java", "-jar", "target/ConjurJavaClient-1.0-SNAPSHOT-with-dependencies.jar"]
```
in light of what this is actually building, is `ConjurJavaClient` the right name?

izgeri commented on java-client-api-tutorial/assets/JavaClient.java:
```diff
@@ -0,0 +1,62 @@
+package com.cyberark.example;
```
is `JavaClient` the right name for this file?

izgeri commented on java-client-api-tutorial/assets/JavaClient.java:
```diff
@@ -0,0 +1,62 @@
+    {
+      System.setProperty("CONJUR_AUTHN_LOGIN", System.getenv("CONJUR_AUTHN_LOGIN"));
+    }
+    System.setProperty("CONJUR_APPLIANCE_URL", System.getenv("CONJUR_APPLIANCE_URL"));
```
this file could benefit from comments explaining what the purpose of the lines of code are

izgeri commented on java-client-api-tutorial/assets/JavaClient.java:
```diff
@@ -0,0 +1,62 @@
+      System.out.println("Exception:" + e);
+      return;
+    }
+    System.out.println("Create Conjur API Instance");
```
```suggestion
    System.out.println("Connect to Conjur server");
```

izgeri commented on java-client-api-tutorial/assets/build.sh:
```diff
@@ -0,0 +1,21 @@
+
+validate_app docker
+
+echo "Creating docker image of Conjur Java Client Example"
```
```suggestion
echo "Building Example Java App"
```

izgeri commented on java-client-api-tutorial/assets/build.sh:
```diff
@@ -0,0 +1,21 @@
+validate_app docker
+
+echo "Creating docker image of Conjur Java Client Example"
+docker build -f Dockerfile --build-arg VERSION=2.2.1 -t conjur-java-client .
```
```suggestion
docker build -f Dockerfile --build-arg VERSION=2.2.1 -t example-java-app .
```

izgeri commented on java-client-api-tutorial/assets/build.sh:
```diff
@@ -0,0 +1,21 @@
+
+echo "Creating docker image of Conjur Java Client Example"
+docker build -f Dockerfile --build-arg VERSION=2.2.1 -t conjur-java-client .
+docker tag conjur-java-client:latest java-demo/conjur-java-client:latest
```
```suggestion
```

izgeri commented on java-client-api-tutorial/assets/build.sh:
```diff
@@ -0,0 +1,21 @@
+echo "Creating docker image of Conjur Java Client Example"
+docker build -f Dockerfile --build-arg VERSION=2.2.1 -t conjur-java-client .
+docker tag conjur-java-client:latest java-demo/conjur-java-client:latest
+docker images | grep conjur-java-client
```
```suggestion
docker images | grep example-java-app
```

izgeri commented on java-client-api-tutorial/assets/conjur-java-api-example.yaml:
```diff
@@ -0,0 +1,98 @@
+          - mountPath: /run/conjur
+            name: conjur-access-token
+      - name: my-conjur-java-client
+        image: java-demo/conjur-java-client:latest
```
```suggestion
        image: example-java-app:latest
```

izgeri commented on java-client-api-tutorial/assets/cert_script.sh:
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
```
cert_script isn't a very descriptive name for this ... is this even used? if not, we can just remove it (in fact, anything that's not being used should be removed from this branch)

izgeri commented on java-client-api-tutorial/assets/conjur-java-api-example.yaml:
```diff
@@ -0,0 +1,98 @@
+        name: authenticator
+        env:
+          - name: CONJUR_VERSION
+            value: "5"
```
```suggestion
```

izgeri commented on java-client-api-tutorial/assets/conjur-java-api-example.yaml:
```diff
@@ -0,0 +1,98 @@
+          - name: CONJUR_AUTHN_URL
+            value: https://conjur-oss.java-demo.svc.cluster.local/authn-k8s/demo-authenticator
+          - name: CONJUR_APPLIANCE_URL
+            value: https://conjur-oss.java-demo.svc.cluster.local
```
we don't recommend in practice that conjur is deployed to the same namespace as the app, so can we deploy conjur in its own namespace?

izgeri commented on java-client-api-tutorial/assets/conjur-java-api-example.yaml:
```diff
@@ -0,0 +1,98 @@
+          - mountPath: /run/conjur
+            name: conjur-access-token
+      initContainers:
+        - name: pem-to-truststore
```
do we need this container? is it actually successful in a minikube / non-openshift env?

izgeri commented on java-client-api-tutorial/assets/custom-values.yaml:
```diff
@@ -0,0 +1,15 @@
+  pullPolicy: Always
+
+ssl:
+  hostname: custom.domainname.com
```
should this be here?

izgeri commented on java-client-api-tutorial/assets/docker-compose.yml:
```diff
@@ -0,0 +1,16 @@
+version: '2'
```
any idea what the purpose of this file is?

izgeri commented on java-client-api-tutorial/assets/installer.sh:
```diff
@@ -0,0 +1,233 @@
+#!/bin/bash
```
can we remove this file?

izgeri commented on java-client-api-tutorial/assets/java-client-installer.sh:
```diff
@@ -0,0 +1,145 @@
+#!/bin/bash
```
can we remove this file?

izgeri commented on java-client-api-tutorial/bootstrap.sh:
```diff
@@ -0,0 +1,9 @@
+export AUTHENTICATOR=demo-authenticator
+
+# Sets the deployment name that would be used by the demo Java client
+export DEPLOYMENT_NAME=conjur-java-api-example
```
rather than having these bootstrap vars, can we replace where they're referenced with hardcoded values?

izgeri commented on java-client-api-tutorial/courseBase.sh:
```diff
@@ -0,0 +1,4 @@
+# Install Helm 3 and overwrite Helm 2
+curl -LO https://get.helm.sh/helm-v3.2.1-linux-amd64.tar.gz
```
with this command here, do we need to have the `java-client-api-tutorial/assets/helm-v3.2.1-linux-amd64.tar.gz` file on this branch?

izgeri commented on java-client-api-tutorial/intro.md:
```diff
@@ -0,0 +1,15 @@
+<p align="center">
+  <img src="assets/cyberark_logo.png">
+</p>
+
```
note: in all your markdown, we still try to have line breaks so the lines aren't longer than ~100 chars. you'll need to edit the files here to reduce line length - lmk if you try it and it makes things look bad in katacoda

izgeri commented on java-client-api-tutorial/intro.md:
```diff
@@ -0,0 +1,15 @@
+
+# Introduction
+
+Welcome to an introduction of the Java client library for the Conjur API. In this tutorial, we will learn how to set up the Java client library for the Conjur API via a Kubernetes environment.
```
```suggestion
In this tutorial, we'll walk through the process from start to finish of deploying a Java app to Kubernetes that uses Conjur as its secret store.

To accomplish this, we will need to:
- Deploy Conjur to Kubernetes using the Conjur Helm Chart
- Add a Kubernetes application to Conjur policy
- Build our sample Java application that uses the Conjur Java client to communicate with Conjur
- Deploy our sample Java application with the Conjur Kubernetes Authenticator Client sidecar, which handles the Conjur authentication for our app
```

izgeri commented on java-client-api-tutorial/index.json:
```diff
@@ -0,0 +1,125 @@
+{
+  "title": "Conjur Java Client API Tutorial",
+  "description": "The purpose of this demo is to install Conjur on a Kubernetes environment",
```
```suggestion
  "title": "Using Java Applications with Conjur in Kubernetes",
  "description": "Deploy a sample Java app to Kubernetes and configure it to communicate with Conjur",
```

izgeri commented on java-client-api-tutorial/intro.md:
```diff
@@ -0,0 +1,15 @@
+
+Welcome to an introduction of the Java client library for the Conjur API. In this tutorial, we will learn how to set up the Java client library for the Conjur API via a Kubernetes environment.
+
+## What is Conjur?
+Conjur is an open source interface to securely authenticate, control and audit non-human access across tools, applications, containers, and cloud environments via robust secrets management.
+
+*If you have never used Conjur before, we highly recommend that you [explore Conjur](https://www.conjur.org/) and read our documentation before starting this tutorial.*
+
+## What is the Java client library for the Conjur API?
+The Java client library provides a robust programmatic interface to a Conjur server from within your Java project. Use the Java SDK to authenticate with Conjur and fetch secrets in a secure manner. Integration with Conjur provides a variety of additional benefits including being able to store security policy as code, and automated secret rotation.
```
I'd maybe just drop this section entirely. We can include links to relevant follow-up resources to learn more at the end. If we want more content on this main page, we could include something from step 1 that highlights the steps, but I think it'd be fine without it.

izgeri commented on java-client-api-tutorial/step1.md:
```diff
@@ -0,0 +1,30 @@
+# Welcome to: Using the Java client library to access the Conjur API!
```
```suggestion
# Welcome to: Using Java Applications with Conjur in Kubernetes!
```

izgeri commented on java-client-api-tutorial/step1.md:
```diff
@@ -0,0 +1,30 @@
+# Welcome to: Using the Java client library to access the Conjur API!
+
+Welcome to Conjur's tutorial of how to use our Java client library to access our Conjur API.
```
```suggestion
Welcome to Conjur's tutorial of how to deploy a Java application to Kubernetes and configure it to communicate with Conjur.
```

izgeri commented on java-client-api-tutorial/step1.md:
```diff
@@ -0,0 +1,30 @@
+
+Welcome to Conjur's tutorial of how to use our Java client library to access our Conjur API.
+
+In this tutorial, you will learn how to run a local Java application that connects to the Conjur API given an access token provided by a Kubernetes authenticator sidecar on a local Kubernetes cluster using Minikube, login to the Conjur CLI, authenticate and set policies, and fetch a secret.
```
```suggestion
In this tutorial, you will learn how to deploy a Java application in Kubernetes that is configured to communicate with a Conjur server. The application will be deployed with the Conjur Kubernetes authenticator client sidecar, which manages the Conjur authentication process and ensures the app always has access to a valid Conjur access token. The application uses the Conjur Java client library to fetch a secret.
```

izgeri commented on java-client-api-tutorial/step1.md:
```diff
@@ -0,0 +1,30 @@
+In this tutorial, you will learn how to run a local Java application that connects to the Conjur API given an access token provided by a Kubernetes authenticator sidecar on a local Kubernetes cluster using Minikube, login to the Conjur CLI, authenticate and set policies, and fetch a secret.
+
+## Overview
+Applications and application developers should be incapable of leaking secrets.
+
```
```suggestion
```
I don't think we're achieving this goal with this tutorial, fwiw

izgeri commented on java-client-api-tutorial/step1.md:
```diff
@@ -0,0 +1,30 @@
+## Overview
+Applications and application developers should be incapable of leaking secrets.
+
+In order to achieve this goal, you will play three roles in this tutorial:
```
```suggestion
In this tutorial, you will play three roles:
```

izgeri commented on java-client-api-tutorial/step1.md:
```diff
@@ -0,0 +1,30 @@
+
+In order to achieve this goal, you will play three roles in this tutorial:
+1. An **Infrastructure Engineer**, who will deploy an instance of Conjur
+2. A **Security Admin** who handles secrets, and has sole access to their secrets
```
```suggestion
2. A **Security Admin** who manages application access control, to ensure apps can only access the resources they're supposed to
```

izgeri commented on java-client-api-tutorial/step1.md:
```diff
@@ -0,0 +1,30 @@
+In order to achieve this goal, you will play three roles in this tutorial:
+1. An **Infrastructure Engineer**, who will deploy an instance of Conjur
+2. A **Security Admin** who handles secrets, and has sole access to their secrets
+3. An **Application Developer** who wants to connect to Conjur to access secrets to be used in their application.
```
```suggestion
3. An **Application Developer** who wants to deploy their Java application to Kubernetes, and make sure it has the credentials to connect to the resources it needs
```
you've been an app developer. when you're building your app, is your primary goal ever "connecting to a secrets service"?

izgeri commented on java-client-api-tutorial/step1.md:
```diff
@@ -0,0 +1,30 @@
+Specifically, we will:
+
+**As the Infrastructure Engineer:**
+1. Create an instance of Conjur OSS and a PostgreSQL database
```
```suggestion
1. Deploy Conjur OSS, its PostgreSQL database, and Nginx for TLS termination in a pod in its own namespace
```

izgeri commented on java-client-api-tutorial/step1.md:
```diff
@@ -0,0 +1,30 @@
+1. Create an instance of Conjur OSS and a PostgreSQL database
+
+**As the Security Admin**
+1. Creates the application identity
```
```suggestion
1. Create a new application identity in Conjur
```

izgeri commented on java-client-api-tutorial/step1.md:
```diff
@@ -0,0 +1,30 @@
+
+**As the Security Admin**
+1. Creates the application identity
+2. Provides the developer with instructions on how to authenticate
+3. Stores the secrets
```
```suggestion
2. Stores the secrets the application will need, and ensure the app will be able to access them
3. Provides the developer with Conjur configuration instructions for their application
```

izgeri commented on java-client-api-tutorial/step1.md:
```diff
@@ -0,0 +1,30 @@
+3. Stores the secrets
+
+**As the Application Developer:**
+1. Create an application to use the token and Java Client to retrieve the given secrets
+2. Update their application manifest
```
```suggestion
1. Update the application manifest to include the Conjur authenticator client sidecar and a volume mount for the Conjur access token
2. Update the application to use the Java client library and the sidecar-supplied Conjur access token to retrieve the credentials the app needs from Conjur
3. Deploy the application, and verify it's up and running!
```

izgeri commented:
we've done a lot of back and forth work on this tutorial, and much thanks to @mdodell for sticking with me on it. it looks like it's going to be hard to trim it down to a reasonable size, so we're going to need to rethink the tutorial we want to write and close this PR for now

thanks again to @mdodell for working so hard on this!!
